west/cmd/git.py

```diff
@@ -0,0 +1,255 @@
+# Copyright (c) 2018 Open Source Foundries Limited.
+#
+# SPDX-License-Identifier: Apache-2.0
+
+'''west git commands'''
+
+import argparse
+import collections
+import os
+import subprocess
+import sys
+
+import pykwalify.core
+import yaml
+
+from . import WestCommand
+from .. import log
+
+
+class Sync(WestCommand):
+    def __init__(self):
+        super().__init__(
+            'sync',
+            'Clone/update all Git repositories specified in the manifest file')
+
+    def do_add_parser(self, parser_adder):
+        return _add_common_git_flags(parser_adder, self)
+
+    def do_run(self, args, user_args):
+        projects = _all_projects(args)
+
+        # TODO: Error checking (if _git(...).returncode != 0: ...)
+
+        for project in projects:
+            if not os.path.exists(project.path):
+                _git_top(project, 'clone -b (branch) (url)/(name) (path)')
+            else:
+                # Fetch first to make sure the project's branch exists. It
```

Comment:
So the thinking here seems to be that we can assume that local branches with the same names track remote branches named in the manifest. As you've noticed, this doesn't play well if the manifest names SHAs (or I guess tags) in its revision fields.

That kind of makes the way repo behaves (as I understand it) seem at least not crazy:

- if detached head, synchronize to latest manifest version
- if there is a local branch checked out, rebase on top of the new manifest branch

Updating a detached HEAD to another detached HEAD seems fine.

I've said before that I think rebase by default is counter-intuitive and not a good idea for West, though -- in my experience teaching people Repo, this is one of the most common frustrations and stumbling blocks, to the point that I usually teach `repo sync -d` first.

If we want to avoid that for West, I think these are the potentially reasonable options if there is a local branch:

1. act like git pull (fetch + merge)
2. act like git fetch (fetch only)
3. scream and die unless a `--rebase` or `--local-branch-strategy={rebase,detach,merge}` or so flag was given
4. always fetch and detach HEAD (unless overridden by a flag)

Am I missing any?

I'm leaning against 1, because it encourages people to submit bad pull requests with upstream merges inside, and 2, because `west sync` ought to, well, synchronize with upstream whenever possible.

The other two seem potentially OK to me.

Thoughts?

Reply:
> 1. act like git pull (fetch + merge)

Would need to check out the branch first in that case (and there would have to
be a local branch). The current version does that, and then switches back to
the previous position. That feels a bit wonky though, but maybe it could work
alright in practice.

> 2. act like git fetch (fetch only)

Would want some way to rebase local work on top of new upstream stuff, so would
probably need additional commands in that case. Wonder if that might be too
complex.

> 3. scream and die unless a --rebase or --local-branch-strategy={rebase,detach,merge} or so flag was given
> 4. always fetch and detach HEAD (unless overridden by a flag)

IMO, it's nice if sync (by default at least) brings in upstream work while also
preserving local changes. I suspect that's what most people would expect.

Keeping a local branch gets a bit tricky when you can have SHAs. I wonder if
there might be gotchas related to changing the revision in the manifest file
too (e.g. to a different branch).

Always just doing a 'git rebase <revision>' is conceptually simplest at least
(and will work with a detached HEAD as well).

The implementation gets a bit tricky though, because e.g. 'git status' won't
know what the upstream is, and so won't tell you stuff like how many commits
behind you are. Gotta implement that stuff manually I think.

Reply:
The only "robust" way I can think of to do it with a non-detached HEAD would be to have a single local branch (e.g. `master`), and updating it against the revision from manifest (ignoring the branch name there).

That'd give you e.g. `git status` for free, but only if the revision from the manifest is a branch name (plus you'd have to update which branch the local branch tracks if the manifest is changed to use a different branch), so might have to implement that stuff manually anyway.

When the revision is an SHA, I don't think the local branch would give you much, beyond avoiding people having to see "detached HEAD" when they run `git status` manually.

Reply:
Hmm... thought of another thing that might work. Wonder if it's too magic though.

Could keep a ref (branch) that's always updated to point to the revision specified in the manifest, and that the user isn't expected to update (could put it in `upstream/revision` or the like, to make it clearer that it's "internal", though it'd be nice to explain it still).

That branch could then be set as the upstream of the local branch, and `git status` and the like would automatically give sensible information (might get confusing for `git push` and the like though).

Might be a horrible idea for reasons I haven't thought of yet though. It's something that's not a detached HEAD at least. :P

Reply:
Seems kinda neat in that you will always have an easy reference for the upstream revision at least, without having to go to the manifest.

Could have `west fetch` update the `upstream/revision` refs, and `west sync` rebase the local branch on top of it, if you wanted to split it up.

Reply:
>> 1. act like git pull (fetch + merge)

> Would need to check out the branch first in that case (and there would have to
> be a local branch).

Right -- note that this list was prefixed with (emphasis added): "I
think these are the potentially reasonable options **if there is a
local branch**". Since we're discussing what to do in case of a local
branch (instead of detached HEAD), it's a safe assumption that there
is one :).

> The current version does that, and then switches back to
> the previous position. That feels a bit wonky though, but maybe it could work
> alright in practice.

Sorry, I don't follow what you're saying. I'm thinking this option 1
would be something like:

```
git fetch <manifest-remote>/<manifest-revision>
git merge FETCH_HEAD
```

That doesn't require switching back to a previous position.

Note that I'm against this option as it encourages invalid pull
request submissions (i.e. PRs which include "merge upstream into my
feature branch" commits).

>> 2. act like git fetch (fetch only)
> Would want some way to rebase local work on top of new upstream stuff, so would
> probably need additional commands in that case. Wonder if that might be too
> complex.

Sorry, I don't follow this either. Why would we *need* a separate
rebase command, when there is already `git rebase`?

We haven't discussed the "submit a pull request" or "fetch a pull
request" commands, but updating local branches to rebase against
upstream might better belong as part of those.

>> 3. scream and die unless a --rebase or --local-branch-strategy={rebase,detach,merge} or so flag was given
>> 4. always fetch and detach HEAD (unless overridden by a flag)
> IMO, it's nice if sync (by default at least) brings in upstream work while also
> preserving local changes. I suspect that's what most people would expect.

> Keeping a local branch gets a bit tricky when you can have SHAs. I wonder if
> there might be gotchas related to changing the revision in the manifest file
> too (e.g. to a different branch).

> Always just doing a git rebase <revision> is conceptually simplest at least
> (and will work with a detached HEAD as well).

I'm confused about what you're saying here. Do think I've missed any
options? Do you lean towards or against any of the four I've listed
above?

> The implementation gets a bit tricky though, because e.g. git status won't
> know what the upstream is, and so won't tell you stuff like how many commits
> behind you are. Gotta implement that stuff manually I think. Would have that
> problem with an SHA regardless though...

Do you want to play around with `repo start`, `repo sync`, and `repo
status` to get a baseline idea of how it works there? That might help
as a starting point, just seeing how at least one other tool works.

The way `repo start` works is that it creates a local branch that
tracks `<manifest-remote>/<manifest-revision>`. So `repo start test
<project>` will create a `test` branch with the following in the
project's .git/config:

```
[branch "test"]
	remote = osf-dev
	merge = refs/heads/master
```

Then `repo sync` will rebase on top of there. And `status` always
knows the remote, so you can do things like:

```
$ repo info -o <project>
Manifest branch: master
Manifest merge branch: refs/heads/master
Manifest groups: all,-notdefault
----------------------------

Projects Overview
<project>
* test                              ( 1 commit, Mon Aug 13 13:13:35 2018 -0500)
                                      - 5136aa26 add test commit
```